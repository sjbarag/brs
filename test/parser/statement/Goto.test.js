const brs = require("brs");
const { token, identifier, EOF } = require("../ParserTests");
const { Lexeme } = brs.lexer;

describe("parser goto statements", () => {
    it("parses standalone statement properly", () => {
        let { statements, errors } = brs.parser.Parser.parse([
            token(Lexeme.Goto, "goto"),
            identifier("SomeLabel"),
            EOF,
        ]);
        expect(errors.length).toEqual(0);
        expect({ errors, statements }).toMatchSnapshot();
    });

    it("detects labels", () => {
        let { statements, errors } = brs.parser.Parser.parse([
            identifier("SomeLabel"),
            token(Lexeme.Colon, ":"),
            EOF,
        ]);
        expect(errors.length).toEqual(0);
        expect(statements).toMatchSnapshot();
    });

    it("allows multiple goto statements on one line", () => {
        let { tokens } = brs.lexer.Lexer.scan(`
            sub Main()
                'multiple goto statements on one line
                goto myLabel : goto myLabel 
                myLabel:
            end sub
        `);
        let { statements, errors } = brs.parser.Parser.parse(tokens);
        expect(errors.length).toEqual(0);
        expect(statements).toMatchSnapshot();
    });
});
