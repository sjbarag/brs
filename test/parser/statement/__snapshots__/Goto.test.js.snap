// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parser goto statements allows multiple goto statements on one line 1`] = `
Array [
  Function {
    "func": Function {
      "body": Block {
        "location": Object {
          "end": Object {
            "column": 12,
            "line": 6,
          },
          "file": "",
          "start": Object {
            "column": 22,
            "line": 2,
          },
        },
        "statements": Array [
          Goto {
            "tokens": Object {
              "goto": Object {
                "isReserved": true,
                "kind": "Goto",
                "literal": undefined,
                "location": Object {
                  "end": Object {
                    "column": 20,
                    "line": 4,
                  },
                  "file": "",
                  "start": Object {
                    "column": 16,
                    "line": 4,
                  },
                },
                "text": "goto",
              },
              "label": Object {
                "isReserved": false,
                "kind": "Identifier",
                "literal": undefined,
                "location": Object {
                  "end": Object {
                    "column": 28,
                    "line": 4,
                  },
                  "file": "",
                  "start": Object {
                    "column": 21,
                    "line": 4,
                  },
                },
                "text": "myLabel",
              },
            },
            "type": "Goto",
          },
          Goto {
            "tokens": Object {
              "goto": Object {
                "isReserved": true,
                "kind": "Goto",
                "literal": undefined,
                "location": Object {
                  "end": Object {
                    "column": 35,
                    "line": 4,
                  },
                  "file": "",
                  "start": Object {
                    "column": 31,
                    "line": 4,
                  },
                },
                "text": "goto",
              },
              "label": Object {
                "isReserved": false,
                "kind": "Identifier",
                "literal": undefined,
                "location": Object {
                  "end": Object {
                    "column": 43,
                    "line": 4,
                  },
                  "file": "",
                  "start": Object {
                    "column": 36,
                    "line": 4,
                  },
                },
                "text": "myLabel",
              },
            },
            "type": "Goto",
          },
          Label {
            "tokens": Object {
              "colon": Object {
                "isReserved": false,
                "kind": "Colon",
                "literal": undefined,
                "location": Object {
                  "end": Object {
                    "column": 24,
                    "line": 5,
                  },
                  "file": "",
                  "start": Object {
                    "column": 23,
                    "line": 5,
                  },
                },
                "text": ":",
              },
              "identifier": Object {
                "isReserved": false,
                "kind": "Identifier",
                "literal": undefined,
                "location": Object {
                  "end": Object {
                    "column": 23,
                    "line": 5,
                  },
                  "file": "",
                  "start": Object {
                    "column": 16,
                    "line": 5,
                  },
                },
                "text": "myLabel",
              },
            },
            "type": "Label",
          },
        ],
        "type": "Block",
      },
      "endKeyword": Object {
        "isReserved": false,
        "kind": "EndSub",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": 19,
            "line": 6,
          },
          "file": "",
          "start": Object {
            "column": 12,
            "line": 6,
          },
        },
        "text": "end sub",
      },
      "keyword": Object {
        "isReserved": true,
        "kind": "Sub",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": 15,
            "line": 2,
          },
          "file": "",
          "start": Object {
            "column": 12,
            "line": 2,
          },
        },
        "text": "sub",
      },
      "parameters": Array [],
      "returns": 11,
      "type": "Expr_Function",
    },
    "name": Object {
      "isReserved": false,
      "kind": "Identifier",
      "literal": undefined,
      "location": Object {
        "end": Object {
          "column": 20,
          "line": 2,
        },
        "file": "",
        "start": Object {
          "column": 16,
          "line": 2,
        },
      },
      "text": "Main",
    },
    "type": "Stmt_Function",
  },
]
`;

exports[`parser goto statements detects labels 1`] = `
Array [
  Label {
    "tokens": Object {
      "colon": Object {
        "isReserved": false,
        "kind": "Colon",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": ":",
      },
      "identifier": Object {
        "isReserved": false,
        "kind": "Identifier",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": "SomeLabel",
      },
    },
    "type": "Label",
  },
]
`;

exports[`parser goto statements parses standalone statement properly 1`] = `
Object {
  "errors": Array [],
  "statements": Array [
    Goto {
      "tokens": Object {
        "goto": Object {
          "isReserved": true,
          "kind": "Goto",
          "literal": undefined,
          "location": Object {
            "end": Object {
              "column": -9,
              "line": -9,
            },
            "start": Object {
              "column": -9,
              "line": -9,
            },
          },
          "text": "goto",
        },
        "label": Object {
          "isReserved": false,
          "kind": "Identifier",
          "literal": undefined,
          "location": Object {
            "end": Object {
              "column": -9,
              "line": -9,
            },
            "start": Object {
              "column": -9,
              "line": -9,
            },
          },
          "text": "SomeLabel",
        },
      },
      "type": "Goto",
    },
  ],
}
`;
